pkg/virt-handler/mount-manager/utils/recorder.go

```diff
@@ -0,0 +1,343 @@
+	ContainerDiskMountStates     = "container-disk-mount-state"
+	HotpluggedVolumesMountStates = "hotplug-volume-mount-state"
+)
+
+type vmiMountTargetRecordForCache struct {
+	MountTargetEntries []MountTargetEntry `json:"mountTargetEntries"`
+	UsesSafePaths      bool               `json:"usesSafePaths"`
+}
+
+func readRecordFile(recordFile string) ([]MountTargetEntry, bool, error) {
+	record := vmiMountTargetRecordForCache{}
+	// #nosec No risk for path injection. Using static base and cleaned filename
+	bytes, err := os.ReadFile(recordFile)
+	if err != nil {
+		return []MountTargetEntry{}, false, err
+	}
+	err = json.Unmarshal(bytes, &record)
+	if err != nil {
+		return []MountTargetEntry{}, false, err
+	}
+	return record.MountTargetEntries, record.UsesSafePaths, nil
+}
+
+func (m *mounter) ReadRecordFiles(uid string) (*VMIMountTargetRecord, bool, error) {
+	record := &VMIMountTargetRecord{}
+	var useSafepathsCd, useSafepathsHp bool
+	// Read the container disks entries from the filesystem
+	// if not there, see if record is on disk, this can happen if virt-handler restarts
+	recordFile := filepath.Join(m.mountStateDir, ContainerDiskMountStates, filepath.Clean(uid))
+	existsCds, err := diskutils.FileExists(recordFile)
+	if err != nil {
+		return nil, false, err
+	}
+	if existsCds {
+		record.ContainerDisks, useSafepathsCd, err = readRecordFile(recordFile)
+		if err != nil {
+			return nil, false, err
+		}
+	}
+
+	// Read hotplugged volumes entries from the filesystem
+	recordFile = filepath.Join(m.mountStateDir, HotpluggedVolumesMountStates, filepath.Clean(uid))
+	if err != nil {
+		return nil, false, err
+	}
+	existsHps, err := diskutils.FileExists(recordFile)
+	if err != nil {
+		return nil, false, err
+	}
+	if existsHps {
+		record.HotpluggedVolumes, useSafepathsHp, err = readRecordFile(recordFile)
+		if err != nil {
+			return nil, false, err
+		}
+	}
+	record.UsesSafePaths = useSafepathsHp && useSafepathsCd
+	return record, existsCds || existsHps, nil
+}
+
+func writeRecordFile(recordFile string, record []MountTargetEntry) error {
+	r := vmiMountTargetRecordForCache{
+		MountTargetEntries: record,
+		// XXX: backward compatibility for old unresolved paths, can be removed in July 2023
+		// After a one-time convert and persist, old records are safe too.
+		UsesSafePaths: true,
+	}
+	bytes, err := json.Marshal(r)
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(filepath.Dir(recordFile), 0750)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(recordFile, bytes, 0600)
+}
+
+func (m *mounter) WriteRecordFiles(uid string, record *VMIMountTargetRecord) error {
+	errCd := writeRecordFile(filepath.Join(m.mountStateDir, ContainerDiskMountStates, uid), record.ContainerDisks)
+	errHp := writeRecordFile(filepath.Join(m.mountStateDir, HotpluggedVolumesMountStates, uid), record.HotpluggedVolumes)
+	return wrapErrors(errCd, errHp)
+}
+
+type RecordEntry int
+
+const (
+	CONTAINERDISKS_ENTRY RecordEntry = iota
+	HOTPLUGGEDVOLUMES_ENTRY
```

Comment:
Is there any need to combine several entry types within the `MountRecorder`?  Why not making it completely unaware of container- or hotplug disks and let it just store 'abstract' state?

I see that in the mount manager you want to use a single instance for all the cases:

```go
	mountRecorder := mountutils.NewMountRecorder(virtPrivateDir)
	return &mountManager{
		containerDiskMounter: container_disk.NewMounter(podIsolationDetector, clusterConfig, mountRecorder),
		hotplugVolumeMounter: hotplug_volume.NewVolumeMounter(mountRecorder, kubeletPodsDir),
	}
```

Is there any reason why we cannot have individual recorders for each type of disks. E.g. smth like this:

```go
	mountRecorder := mountutils.NewMountRecorder(virtPrivateDir)
	return &mountManager{
		containerDiskMounter: container_disk.NewMounter(podIsolationDetector, clusterConfig, mountutils.NewMountRecorder(virtPrivateDir, "container-disks-state")),
		hotplugVolumeMounter: hotplug_volume.NewVolumeMounter(mountutils.NewMountRecorder(virtPrivateDir, "hotplug-disks-state"), kubeletPodsDir),
	}
```

Reply:
hi @vasiliy-ul, if I understand your question correctly, it is similar to [Alexander's comment](https://github.com/kubevirt/kubevirt/pull/8317#discussion_r1020419617). We need to distinguish the type of disk because hotplugged disks could be deleted while VM still running and this isn't the case for container disks.

Reply:
Right, seems I have the same concern as Alexander. Sorry for not having read through all the previous comments :)

Though I still fail to understand why we need to distinguish between container- and hotplug disks within the recorder.... The recorder in my understanding should be unaware of any logic specific to the disks. It just keeps the state in a file. You can simply use separate instances of MountRecorder as I mentioned above and that way you should not worry about the right moment for deletion. I think the deletion logic should be (... and likely is already) implemented in the level above where the recorder is actaully used.

Reply:
@vasiliy-ul I agree with you, the recorder shouldn't make the difference between the disks type, but unfortunately, it isn't the case as the code evolved. With a single instance of the mount manager, I need to distinguish the 2 types otherwise, when the hotplugged volumes are deleted then also the containerdisks entries are removed. The main reason for this is that I tried to stay close to the original code that was basically the same in 2 functions for hotplugged and container disks.

If I do what you suggested, using 2 instances of the mount manager, then I see fewer benefits in using a mount interface. I could simply only refactor the recorder and not the mount. This mount manager should abstract all the mount operations and keep them in a single place.

I have failed to find a nice way to solve the clean up. Does this make sense to you?

Reply:
IOW, my original idea was that from virt-handler (`vm.go`), we could simply call `mount.Mount(vmi)`, and then based on the VMI description all disk types are handled there. If I split it in 2 or mount manager, then we missed this abstraction in the main virt-handler code.

Reply:
> when the hotplugged volumes are deleted then also the containerdisks entries are removed

Hm... If I understand correctly, the MountRecorder is an object that keeps track of the existing mounts using a runtime map (i.e. in memory) and also maintains a file cache for the case if virt-handler is restarted, right? Then if we track the mounts for container- and hotplug disks with different MountRecorder instances, then those will have independent maps and cache files. How can a deletion of a hotplug record delete smth in the containerdisk instance?

I quickly sketched a commit with a refactor https://github.com/vasiliy-ul/kubevirt/commit/ac1c2b6f88438f9c26f62a0b0be17728a6c8dfaa# I haven't tested it, just made it buildable. But I think it should work as expected.... unless I fail to see something obvious in the picture. Perhaps you could take a look?

Reply:
>  If I split it in 2 or mount manager, then we missed this abstraction in the main virt-handler code.

Maybe I expressed myself not very clear... My proposal was to make two recorders within one mount manager. Anyway, this should be reflected in the commit linked above.

Reply:
@vasiliy-ul thanks! The code helped! I somehow understood to use 2 instances of the mount manager and not the recorder. Anyway, that's a great idea :D, it really simplifies the recorder. At first look, it should work. I'll try to adapt the code!

Reply:
@vasiliy-ul one problem with your suggestion here is that we are changing the json of the record entry. In the new version you are replacing `hotpluggedDisks` or `containerDisks`  with `mountTargetEntries. It won't work when we upgrade from a version that doesn't have this naming conversion. I already faced a similar issue when I tried to merge the entries into a single file: https://github.com/kubevirt/kubevirt/pull/8317#issuecomment-1303093512

Reply:
We could add some legacy struct for reading the new version of the json

Reply:
apparently the tests for upgrading didn't fail :thinking: I might have missed something

Reply:
> we are changing the json of the record entry

Do we? The new struct seems compatible:

```go
type MountTargetEntry struct {
	TargetFile string `json:"targetFile"`
	SocketFile string `json:"socketFile,omitempty"`
}

type vmiMountTargetRecord struct {
	MountTargetEntries []MountTargetEntry `json:"mountTargetEntries"`
	UsesSafePaths      bool               `json:"usesSafePaths"`
}
```

Exactly the same json fields are used for container- and hotplug disks in the existing implementation. The only thing is that `SocketFile` is not needed for hotplug disks. But I guess that should not cause troubles as long as we mark it with `omitempty`.

Reply:
@vasiliy-ul yes you are right! I was thinking that we had "containeddisks" or "hoplugged volumes" hardcoded in the JSON but this isn't the case :) I just checked